httpone.go

```diff
@@ -0,0 +1,197 @@
+	s.challMu.RLock()
+	defer s.challMu.RUnlock()
+	content, present := s.httpOne[token]
+	return content, present
+}
+
+// AddHTTPRedirect adds a redirect for the given path to the given URL.
+func (s *ChallSrv) AddHTTPRedirect(path, targetURL string) {
+	s.challMu.Lock()
+	defer s.challMu.Unlock()
+	s.redirects[path] = targetURL
+}
+
+// DeletedHTTPRedirect deletes a redirect for the given path.
+func (s *ChallSrv) DeleteHTTPRedirect(path string) {
+	s.challMu.Lock()
+	defer s.challMu.Unlock()
+	if _, ok := s.redirects[path]; ok {
+		delete(s.redirects, path)
+	}
+}
+
+// GetHTTPRedirect returns the redirect target for the given path
+// (if it exists) and a true bool. If the path does not have a redirect target
+// then an empty string and a false bool are returned.
+func (s *ChallSrv) GetHTTPRedirect(path string) (string, bool) {
+	s.challMu.RLock()
+	defer s.challMu.RUnlock()
+	targetURL, present := s.redirects[path]
+	return targetURL, present
+}
+
+// ServeHTTP handles an HTTP request. If the request path has the ACME HTTP-01
+// challenge well known prefix as a prefix and the token specified is known,
+// then the challenge response contents are returned.
+func (s *ChallSrv) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	requestPath := r.URL.Path
+
+	// If the request was not over HTTPS and we have a redirect, serve it.
+	// Redirects are ignored over HTTPS so we can easily do an HTTP->HTTPS
+	// redirect for a token path without creating a loop.
+	if redirectTarget, found := s.GetHTTPRedirect(requestPath); found && r.TLS == nil {
+		http.Redirect(w, r, redirectTarget, http.StatusFound)
+		return
+	}
+
+	if strings.HasPrefix(requestPath, wellKnownPath) {
+		token := requestPath[len(wellKnownPath):]
+		if auth, found := s.GetHTTPOneChallenge(token); found {
+			fmt.Fprintf(w, "%s", auth)
+		}
+	}
+}
+
+// challHTTPServer is a *http.Server that has a Shutdown() func that doesn't
+// take a context argument. This lets us treat the HTTP server the same as the
+// DNS-01 servers (which use a `dns.Server` that has `Shutdown()` with no
+// context arg) by having an http.Server that implements the challengeServer
+// interface.
+type challHTTPServer struct {
+	*http.Server
+}
+
+// ListenAndServe for a challHTTPServer will call the underlying http.Server's
+// ListenAndServeTLS if the server has a non-nil TLSConfig, otherwise it will
+// use the underlying http.Server's ListenAndServe(). This allos for
+// a challHTTPServer to be both a normal HTTP based HTTP-01 challenge response
+// server in one configuration (nil TLSConfig) and an HTTPS based HTTP-01
+// challenge response server useful for redirect targets in another
+// configuration.
+func (c challHTTPServer) ListenAndServe() error {
+	// If the server has a TLSConfig then use ListenAndServeTLS with empty
+	// certfile/keyfile arguments. This will use the self-signed certificate we
+	// issued when constructing the TLSConfig in `httpOneServer`.
```

Comment:
We actually issued the self-signed certificate at init time (via the var assignment above). Also, it might be clearer here to say:

"This will use the certificate and key from TLSConfig"

Then, below at line 179-184 we can see that we're using the self-signed certificate generated at startup.

Duplicating the comment both here and there is suboptimal because it increases the chance that one comment will become out of date.